### PR TITLE
ext-auth: implement failure_mode_allow

### PR DESCRIPTION
**Pull request by hanxiantao**
fixes: #49 
集成测试结果如下：
image

### CODE REVIEW

File: types/plugins/ext_auth/config.proto [hunk 1]
```diff
@@ -61,18 +81,18 @@ message AuthorizationRequest {
   // Sets a list of headers that will be included to the request to authorization service. Note that
   // client request of the same key will be overridden.
   repeated api.v1.HeaderValue headers_to_add = 1
-      [(validate.rules).repeated = {ignore_empty: true, min_items: 1}];
+  [(validate.rules).repeated = {ignore_empty: true, min_items: 1}];
```
Comment (spacewander) [inline types/plugins/ext_auth/config.proto:84]: It seems you are using a different clang-format? You can install this one:
https://github.com/mosn/htnn/blob/0be21bc69650e3872672555ef76d843856a2ed98/tools/Dockerfile.dev#L70

File: plugins/tests/integration/ext_auth_test.go [hunk 2]
```diff
@@ -79,6 +79,20 @@ func TestExtAuth(t *testing.T) {
...
+					"url":           "http://127.0.0.1:2023/ext_auth",
+					"statusOnError": 403,
+				},
+				"failure_mode_allow": true,
+			}),
+			run: func(t *testing.T) {
+				resp, _ := dp.Post("/echo", nil, strings.NewReader("any"))
+				assert.Equal(t, 200, resp.StatusCode)
+			},
+		},
```
Comment (spacewander) [inline plugins/tests/integration/ext_auth_test.go:95]: Would you add a test to cover "failure_mode_allow_header_add"?

File: types/plugins/ext_auth/config.proto [hunk 3]
```diff
@@ -32,6 +32,26 @@ message Config {
...
     HttpService http_service = 1;
   }
+
+  //  Changes filter's behaviour on errors:
+  //
+  //  1. When set to true, the filter will ``accept`` client request even if the communication with
+  //  the authorization service has failed, or if the authorization service has returned a HTTP 5xx
+  //  error.
+  //
+  //  2. When set to false, ext-authz will ``reject`` client requests and return a ``Forbidden``
```
Comment (spacewander) [inline types/plugins/ext_auth/config.proto:42]:
```suggestion
  //  2. When set to false, ext-auth will ``reject`` client requests and return a ``Forbidden``
```

File: types/plugins/ext_auth/config.proto [hunk 4]
```diff
@@ -32,6 +32,26 @@ message Config {
...
+  //  response if the communication with the authorization service has failed, or if the
+  //  authorization service has returned a HTTP 5xx error.
+  //
+  // Note that errors can be ``always`` tracked in the :ref:`stats
```
Comment (spacewander) [inline types/plugins/ext_auth/config.proto:46]: We can remove this part. It's not true in the Go implementation.

File: plugins/tests/integration/ext_auth_test.go (hunk 2 above)
Comment (hanxiantao) [reply plugins/tests/integration/ext_auth_test.go:95]: 咨询下这个单元测试本地跑的时候，需要启动项目吗，有没有什么脚本，昨天搞的时候没找到

**Review by spacewander — COMMENTED** [review]
Let's update the doc: https://github.com/mosn/htnn/blob/main/site/content/zh-hans/docs/reference/plugins/ext_auth.md and its English version. We have tool to generate prompt to translate the doc: https://github.com/mosn/htnn/tree/main/site#cmdtranslator. Or you can let me do it for you.

File: plugins/tests/integration/ext_auth_test.go (hunk 2 above)
Comment (hanxiantao) [reply plugins/tests/integration/ext_auth_test.go:95]: > Would you add a test to cover "failure_mode_allow_header_add"?

When running unit tests locally, do I need to start the project? Are there any scripts available? I haven't found any yet.

**Comment by hanxiantao** [comment]
> Let's update the doc: https://github.com/mosn/htnn/blob/main/site/content/zh-hans/docs/reference/plugins/ext_auth.md and its English version. We have tool to generate prompt to translate the doc: https://github.com/mosn/htnn/tree/main/site#cmdtranslator. Or you can let me do it for you.

I will address this along with the previous questions later.

File: plugins/tests/integration/ext_auth_test.go (hunk 2 above)
Comment (spacewander) [reply plugins/tests/integration/ext_auth_test.go:95]: 这里可以先切到 `./plugins` 目录下， `make build-test-so && make integration-test` 来运行测试。
https://github.com/mosn/htnn/blob/0be21bc69650e3872672555ef76d843856a2ed98/plugins/Makefile#L55

不过这个是执行全量集成测试的，需要 make start-service 准备一些外部服务。本地跑的话，可以只跑这个插件的测试，无需准备额外服务。直接 make build-test-so，然后 go test -v ./tests/integration/ -run TestExtAuth 就行了。

我看文档里面没有记录这一点：https://github.com/mosn/htnn/blob/main/site/content/zh-hans/docs/developer-guide/plugin_development.md
你后面能不能再提个 PR 更新下文档？多谢！

Comment (hanxiantao) [reply plugins/tests/integration/ext_auth_test.go:95]: 好的，我本地验证没问题之后，会再更新下文档

Comment (spacewander) [reply plugins/tests/integration/ext_auth_test.go:95]: 这个就是我前面提到的测试方法。如果插件开发文档里面没有提及，可以在插件开发文档里加个链接。

File: site/content/en/docs/reference/plugins/ext_auth.md [hunk 5]
```diff
@@ -15,9 +15,11 @@ The `extAuth` plugin sends an authorization request to an authorization service
 
 ## Configuration
 
-| Name        | Type        | Required | Validation | Description |
-| ----------- | ----------- | -------- | ---------- | ----------- |
-| httpService | HttpService | True     |            |             |
+| Name                          | Type        | Required | Validation | Description                                                  |
+| ----------------------------- | ----------- | -------- | ---------- | ------------------------------------------------------------ |
+| httpService                   | HttpService | True     |            |                                                              |
+| failure_mode_allow            | bool        | False    |            | Default is "false".When set to true, the filter will “accept“ client request even if the communication with the authorization service has failed, or if the authorization service has returned a HTTP 5xx |
```
Comment (spacewander) [inline site/content/en/docs/reference/plugins/ext_auth.md:21]:
```suggestion
| failure_mode_allow            | bool        | False    |            | Default is `false`. When set to true, the filter will "accept" client request even if the communication with the authorization service has failed, or if the authorization service has returned an HTTP 5xx |
```

File: site/content/en/docs/reference/plugins/ext_auth.md [hunk 6]
```diff
@@ -15,9 +15,11 @@ The `extAuth` plugin sends an authorization request to an authorization service
...
+| failure_mode_allow_header_add | bool        | False    |            | Default is "false".When “failure_mode_allow“ and “failure_mode_allow_header_add“ are both set to true, “x-envoy-auth-failure-mode-allowed: true“ will be added to request headers if the communication with the authorization service has failed, or if the authorization service has returned a HTTP 5xx error |
```
Comment (spacewander) [inline site/content/en/docs/reference/plugins/ext_auth.md:22]:
```suggestion
| failure_mode_allow_header_add | bool        | False    |            | Default is `false`. When `failure_mode_allow` and `failure_mode_allow_header_add` are both set to true, "x-envoy-auth-failure-mode-allowed: true" will be added to request headers if the communication with the authorization service has failed, or if the authorization service has returned an HTTP 5xx error |
```

File: site/content/zh-hans/docs/reference/plugins/ext_auth.md [hunk 7]
```diff
@@ -18,6 +18,8 @@ title: Ext Auth
 | 名称          | 类型          | 必选 | 校验规则 | 说明 |
 |--------------|---------------|------|----------|------|
 | httpService | HttpService   | 是   |          |      |
+| failure_mode_allow | bool | 否 | | 默认值false。当设置为true时，即使与授权服务的通信失败，或者授权服务返回了HTTP 5xx错误，过滤器仍会接受客户端请求 |
```
Comment (spacewander) [inline site/content/zh-hans/docs/reference/plugins/ext_auth.md:21]:
```suggestion
| failure_mode_allow | bool | 否 | | 默认为 false。当设置为 true 时，即使与授权服务的通信失败，或者授权服务返回了 HTTP 5xx 错误，过滤器仍会接受客户端请求 |
```

File: site/content/zh-hans/docs/reference/plugins/ext_auth.md [hunk 8]
```diff
@@ -18,6 +18,8 @@ title: Ext Auth
...
+| failure_mode_allow_header_add | bool | 否 | | 默认值false。当`failure_mode_allow`和`failure_mode_allow_header_add`都设置为true时，若与授权服务的通信失败，或授权服务返回了HTTP 5xx错误，那么请求头中将会添加`x-envoy-auth-failure-mode-allowed: true` |
```
Comment (spacewander) [inline site/content/zh-hans/docs/reference/plugins/ext_auth.md:22]:
```suggestion
| failure_mode_allow_header_add | bool | 否 | | 默认为 false。当 `failure_mode_allow` 和 `failure_mode_allow_header_add` 都设置为 true 时，若与授权服务的通信失败，或授权服务返回了 HTTP 5xx 错误，那么请求头中将会添加 `x-envoy-auth-failure-mode-allowed: true` |
```

**Review by spacewander — APPROVED** [review]

**Comment by spacewander** [comment]
@hanxiantao 
Merged, thanks!

**Pull request merged**